hoomd/md/NeighborList.cc

```diff
@@ -1742,6 +1764,253 @@ void NeighborList::updateMemoryMapping()
+    ArrayHandle<unsigned int> h_nlist(getNListArray(), access_location::host, access_mode::read);
+    ArrayHandle<size_t> h_head_list(getHeadList(), access_location::host, access_mode::read);
+
+    auto* pair_list = new std::vector<vec2<uint32_t>>();
+
+    pybind11::capsule delete_when_done(pair_list,
+                                       [](void* f)
+                                       {
+                                           auto data
+                                               = reinterpret_cast<std::vector<vec2<uint32_t>>*>(f);
+                                           delete data;
+                                       });
+
+    for (unsigned int i = 0; i < m_pdata->getN(); i++)
+        {
+        const size_t myHead = h_head_list.data[i];
+        const unsigned int size = h_n_neigh.data[i];
+        for (unsigned int k = 0; k < size; k++)
+            {
+            const unsigned int j = h_nlist.data[myHead + k];
+            // if j is not a ghost, only accept it if i < j
+            if (!third_law && j < m_pdata->getN() && i > j)
+                continue;
+
+            pair_list->push_back(vec2(i, j));
+            }
+        }
+
+    unsigned long shape[] = {pair_list->size(), 2};
+    unsigned long stride[] = {2 * sizeof(uint32_t), sizeof(uint32_t)};
+
+    return pybind11::array_t(shape,                        // shape
+                             stride,                       // stride
+                             (uint32_t*)pair_list->data(), // data pointer
+                             delete_when_done);
+    }
+
+pybind11::object NeighborList::getPairListPython(uint64_t timestep)
+    {
+    compute(timestep);
+
+    bool third_law = getStorageMode() == NeighborList::half;
+
+#ifdef ENABLE_MPI
+    // if we are not the root processor, return None
+    bool root = m_exec_conf->isRoot();
+#endif
+
+    ArrayHandle<unsigned int> h_n_neigh(getNNeighArray(), access_location::host, access_mode::read);
+    ArrayHandle<unsigned int> h_nlist(getNListArray(), access_location::host, access_mode::read);
+    ArrayHandle<size_t> h_head_list(getHeadList(), access_location::host, access_mode::read);
+
+    ArrayHandle<unsigned int> h_tags(m_pdata->getTags(), access_location::host, access_mode::read);
+
+    auto* pair_list = new std::vector<vec2<uint32_t>>();
+
+    for (int i = 0; i < (int)m_pdata->getN(); i++)
+        {
+        unsigned int tag_i = h_tags.data[i];
+        const size_t myHead = h_head_list.data[i];
+        const unsigned int size = h_n_neigh.data[i];
+        for (unsigned int k = 0; k < size; k++)
+            {
+            const unsigned int j = h_nlist.data[myHead + k];
+            const unsigned int tag_j = h_tags.data[j];
+            if ((!third_law || j >= m_pdata->getN()) && tag_i > tag_j)
+                continue;
+
+            pair_list->push_back(vec2(tag_i, tag_j));
+            }
+        }
+
+    // pair_list->shrink_to_fit();
```

Comment:
```suggestion
    pair_list->shrink_to_fit();
```
Why not shrink to fit and save memory utilization? You could also call it in `getLocalPairListPython`.

Reply:
Was unsure whether it should be called, since it is an O(N) copy operation. As a side note, the neighborlist (I'm pretty sure) doesn't retain a count of the total number of neighbors. Maybe as a later change, we could keep track of that and reduce the number of allocations and copies that occur.

Reply:
Though yes, it's good here to shrink and reduce the memory allocation, as it could be quite big.

Reply:
The GPU computes many per-particle neighbor lists in parallel. It would take `O(N log N)` extra work to perform a parallel reduction to compute the total number of neighbors. You are performing `O(N)` work here.

The other disadvantage of `shrink_to_fit` is that you end up storing the oversized list to start with, so the memory requirements are there - even if temporary. The only alternative to that is an O(N) sum *first*, then pre-allocate `pair_list` to the exact correct size and then populate it.

Reply:
You can get the total number of neighbors by reducing `h_n_neigh`. It should be a one-liner with `std::accumulate`. There is an equivalent `thrust::reduce` if you want to do the same on the GPU.